File: apps/backend-node/src/features/notifications/notification.service.ts
import { Injectable } from '@nestjs/common';

import type { AuthUser } from '../../shared/auth/auth-user.model';
import { NotFoundError } from '../../shared/errors';
import type {
  MarkAllNotificationsReadResultDto,
  NotificationDto,
  UnreadNotificationCountDto,
} from './dto';
import { mapNotification } from './mappers/notification.mapper';
import { NotificationRepository } from './notification.repository';

@Injectable()
export class NotificationService {
  constructor(private readonly repository: NotificationRepository) {}

  async listNotifications(user: AuthUser): Promise<NotificationDto[]> {
    const rows = await this.repository.listForUser(user.id);
    return rows.map(mapNotification);
  }

  async getUnreadCount(user: AuthUser): Promise<UnreadNotificationCountDto> {
    const count = await this.repository.countUnread(user.id);
    return { count };
  }

  async markRead(
    user: AuthUser,
    notificationId: string,
  ): Promise<NotificationDto> {
    const notification = await this.repository.markRead(user.id, notificationId);
    if (!notification) {
      throw new NotFoundError('Notification not found');
    }

    return mapNotification(notification);
  }

  async markAllRead(
    user: AuthUser,
  ): Promise<MarkAllNotificationsReadResultDto> {
    const updatedCount = await this.repository.markAllRead(user.id);
    return { updatedCount };
  }
}

File: apps/backend-java/src/main/java/com/housingplatform/notifications/NotificationService.java
package com.housingplatform.notifications;

import com.housingplatform.auth.error.NotFoundException;
import com.housingplatform.auth.model.AuthenticatedUser;
import com.housingplatform.notifications.dto.MarkAllNotificationsReadResult;
import com.housingplatform.notifications.dto.NotificationDto;
import com.housingplatform.notifications.dto.UnreadNotificationCount;
import com.housingplatform.notifications.mapper.NotificationMapper;
import com.housingplatform.persistence.repository.NotificationRepository;
import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class NotificationService {

  private final NotificationRepository notificationRepository;

  public NotificationService(NotificationRepository notificationRepository) {
    this.notificationRepository = notificationRepository;
  }

  @Transactional(readOnly = true)
  public List<NotificationDto> listNotifications(AuthenticatedUser user) {
    return notificationRepository.findTop50ByUserIdOrderByCreatedAtDesc(user.id()).stream()
        .map(NotificationMapper::toDto)
        .toList();
  }

  @Transactional(readOnly = true)
  public UnreadNotificationCount getUnreadCount(AuthenticatedUser user) {
    return new UnreadNotificationCount(
        Math.toIntExact(notificationRepository.countByUserIdAndReadAtIsNull(user.id())));
  }

  @Transactional
  public NotificationDto markRead(AuthenticatedUser user, UUID notificationId) {
    notificationRepository.markRead(user.id(), notificationId);
    var notification =
        notificationRepository
            .findByIdAndUserId(notificationId, user.id())
            .orElseThrow(() -> new NotFoundException("Notification not found"));
    return NotificationMapper.toDto(notification);
  }

  @Transactional
  public MarkAllNotificationsReadResult markAllRead(AuthenticatedUser user) {
    int updatedCount = notificationRepository.markAllRead(user.id());
    return new MarkAllNotificationsReadResult(updatedCount);
  }
}
